### ISSUE TITLE
Out-of-memory for filter not exists

### ISSUE THREAD

**Issue by MPvHarmelen** [comunica/comunica#835]
#### Issue type:

- :bug: Bug 

____
#### Description:

Using the following example, Comunica crashes because it runs out of memory.

```js
const { Store, Parser } = require("n3");
const { newEngine } = require("@comunica/actor-init-sparql-rdfjs");
const data = `
prefix : <https://example.com/>
:m :a ( :a1 :a2 :a3 :a4 :a5 :a6 :a7 :a8 :a9 :a10 :a11 :a12 :a13 :a14 :a15 :a16 :a17 :a18 :a19 :a20 :a21 :a22 :a23 :a24 :a25 :a26 :a27 :a28 :a29 :a30 :a31 :a32 :a33 :a34 :a35 :a36 :a37 :a38 :a39 :a40 :a41 :a42 :a43 :a44 :a45 :a46 :a47 :a48 :a49 :a50 :a51 :a52 :a53 :a54 :a55 :a56 :a57 :a58 :a59 :a60 :a61 :a62 :a63 :a64 :a65 :a66 :a67 :a68 :a69 :a70 :a71 :a72 :a73 :a74 :a75 :a76 :a77 :a78 :a79 :a80 :a81 :a82 :a83 :a84 :a85 :a86 :a87 :a88 :a89) ;
   :b ( :b1 :b2 :b3 :b4 :b5 :b6 :b7 :b8 :b9 :b10 :b11 :b12 :b13 :b14 :b15 :b16 :b17 :b18 :b19 :b20 :b21 :b22 :b23 :b24 :b25 :b26 :b27 :b28 :b29 :b30 :b31 :b32 :b33 :b34 :b35 :b36 :b37 :b38 :b39 :b40 :b41 :b42 :b43 :b44 :b45 :b46 :b47 :b48 :b49 :b50 :b51 :b52 :b53 :b54 :b55 :b56 :b57 :b58 :b59 :b60 :b61 :b62 :b63 :b64 :b65 :b66 :b67 :b68 :b69 :b70 :b71 :b72 :b73 :b74 :b75 :b76 :b77 :b78 :b79 :b80 :b81 :b82 :b83 :b84 :b85 :b86 :b87 :b88 :b89) ;
   :c ( :c1 :c2 :c3 :c4 :c5 :c6 :c7 :c8 :c9 :c10 :c11 :c12 :c13 :c14 :c15 :c16 :c17 :c18 :c19 :c20 :c21 :c22 :c23 :c24 :c25 :c26 :c27 :c28 :c29 :c30 :c31 :c32 :c33 :c34 :c35 :c36 :c37 :c38 :c39 :c40 :c41 :c42 :c43 :c44 :c45 :c46 :c47 :c48 :c49 :c50 :c51 :c52 :c53 :c54 :c55 :c56 :c57 :c58 :c59 :c60 :c61 :c62 :c63 :c64 :c65 :c66 :c67 :c68 :c69 :c70 :c71 :c72 :c73 :c74 :c75 :c76 :c77 :c78 :c79 :c80 :c81 :c82 :c83 :c84 :c85 :c86 :c87 :c88 :c89) ;
   :d ( :d1 :d2 :d3 :d4 :d5 :d6 :d7 :d8 :d9 :d10 :d11 :d12 :d13 :d14 :d15 :d16 :d17 :d18 :d19 :d20 :d21 :d22 :d23 :d24 :d25 :d26 :d27 :d28 :d29 :d30 :d31 :d32 :d33 :d34 :d35 :d36 :d37 :d38 :d39 :d40 :d41 :d42 :d43 :d44 :d45 :d46 :d47 :d48 :d49 :d50 :d51 :d52 :d53 :d54 :d55 :d56 :d57 :d58 :d59 :d60 :d61 :d62 :d63 :d64 :d65 :d66 :d67 :d68 :d69 :d70 :d71 :d72 :d73 :d74 :d75 :d76 :d77 :d78 :d79 :d80 :d81 :d82 :d83 :d84 :d85 :d86 :d87 :d88 :d89) ;
   :e ( :e1 :e2 :e3 :e4 :e5 :e6 :e7 :e8 :e9 :e10 :e11 :e12 :e13 :e14 :e15 :e16 :e17 :e18 :e19 :e20 :e21 :e22 :e23 :e24 :e25 :e26 :e27 :e28 :e29 :e30 :e31 :e32 :e33 :e34 :e35 :e36 :e37 :e38 :e39 :e40 :e41 :e42 :e43 :e44 :e45 :e46 :e47 :e48 :e49 :e50 :e51 :e52 :e53 :e54 :e55 :e56 :e57 :e58 :e59 :e60 :e61 :e62 :e63 :e64 :e65 :e66 :e67 :e68 :e69 :e70 :e71 :e72 :e73 :e74 :e75 :e76 :e77 :e78 :e79 :e80 :e81 :e82 :e83 :e84 :e85 :e86 :e87 :e88 :e89) ;
   :f ( :f1 :f2 :f3 :f4 :f5 :f6 :f7 :f8 :f9 :f10 :f11 :f12 :f13 :f14 :f15 :f16 :f17 :f18 :f19 :f20 :f21 :f22 :f23 :f24 :f25 :f26 :f27 :f28 :f29 :f30 :f31 :f32 :f33 :f34 :f35 :f36 :f37 :f38 :f39 :f40 :f41 :f42 :f43 :f44 :f45 :f46 :f47 :f48 :f49 :f50 :f51 :f52 :f53 :f54 :f55 :f56 :f57 :f58 :f59 :f60 :f61 :f62 :f63 :f64 :f65 :f66 :f67 :f68 :f69 :f70 :f71 :f72 :f73 :f74 :f75 :f76 :f77 :f78 :f79 :f80 :f81 :f82 :f83 :f84 :f85 :f86 :f87 :f88 :f89) ;
   :g ( :g1 :g2 :g3 :g4 :g5 :g6 :g7 :g8 :g9 :g10 :g11 :g12 :g13 :g14 :g15 :g16 :g17 :g18 :g19 :g20 :g21 :g22 :g23 :g24 :g25 :g26 :g27 :g28 :g29 :g30 :g31 :g32 :g33 :g34 :g35 :g36 :g37 :g38 :g39 :g40 :g41 :g42 :g43 :g44 :g45 :g46 :g47 :g48 :g49 :g50 :g51 :g52 :g53 :g54 :g55 :g56 :g57 :g58 :g59 :g60 :g61 :g62 :g63 :g64 :g65 :g66 :g67 :g68 :g69 :g70 :g71 :g72 :g73 :g74 :g75 :g76 :g77 :g78 :g79 :g80 :g81 :g82 :g83 :g84 :g85 :g86 :g87 :g88 :g89) ;
   :h ( :h1 :h2 :h3 :h4 :h5 :h6 :h7 :h8 :h9 :h10 :h11 :h12 :h13 :h14 :h15 :h16 :h17 :h18 :h19 :h20 :h21 :h22 :h23 :h24 :h25 :h26 :h27 :h28 :h29 :h30 :h31 :h32 :h33 :h34 :h35 :h36 :h37 :h38 :h39 :h40 :h41 :h42 :h43 :h44 :h45 :h46 :h47 :h48 :h49 :h50 :h51 :h52 :h53 :h54 :h55 :h56 :h57 :h58 :h59 :h60 :h61 :h62 :h63 :h64 :h65 :h66 :h67 :h68 :h69 :h70 :h71 :h72 :h73 :h74 :h75 :h76 :h77 :h78 :h79 :h80 :h81 :h82 :h83 :h84 :h85 :h86 :h87 :h88 :h89) ;
   :i ( :i1 :i2 :i3 :i4 :i5 :i6 :i7 :i8 :i9 :i10 :i11 :i12 :i13 :i14 :i15 :i16 :i17 :i18 :i19 :i20 :i21 :i22 :i23 :i24 :i25 :i26 :i27 :i28 :i29 :i30 :i31 :i32 :i33 :i34 :i35 :i36 :i37 :i38 :i39 :i40 :i41 :i42 :i43 :i44 :i45 :i46 :i47 :i48 :i49 :i50 :i51 :i52 :i53 :i54 :i55 :i56 :i57 :i58 :i59 :i60 :i61 :i62 :i63 :i64 :i65 :i66 :i67 :i68 :i69 :i70 :i71 :i72 :i73 :i74 :i75 :i76 :i77 :i78 :i79 :i80 :i81 :i82 :i83 :i84 :i85 :i86 :i87 :i88 :i89) ;
   :j ( :j1 :j2 :j3 :j4 :j5 :j6 :j7 :j8 :j9 :j10 :j11 :j12 :j13 :j14 :j15 :j16 :j17 :j18 :j19 :j20 :j21 :j22 :j23 :j24 :j25 :j26 :j27 :j28 :j29 :j30 :j31 :j32 :j33 :j34 :j35 :j36 :j37 :j38 :j39 :j40 :j41 :j42 :j43 :j44 :j45 :j46 :j47 :j48 :j49 :j50 :j51 :j52 :j53 :j54 :j55 :j56 :j57 :j58 :j59 :j60 :j61 :j62 :j63 :j64 :j65 :j66 :j67 :j68 :j69 :j70 :j71 :j72 :j73 :j74 :j75 :j76 :j77 :j78 :j79 :j80 :j81 :j82 :j83 :j84 :j85 :j86 :j87 :j88 :j89) .
`;
const query = `SELECT * WHERE {
    filter not exists { ?list rdf:rest/rdf:first []. }
    ?list rdf:first [].
}`;

async function run() {
  const store = new Store();
  store.addQuads(new Parser().parse(data));
  const engine = newEngine();
  await engine.query(query, { sources: [store] });
}
run();
```

Note that the amount of memory used seems to be proportional to the number of RDF-lists in the data, as well as the size of those lists.

We're reporting to your issue tracker (and not that of N3), because we do not find any memory issues when we first log all calls to `match` using the following source (until the process crashes):
```js
{
  match: (...args) => {
    console.log(JSON.stringify(args));
    return store.match(...args);
  },
}
```
and then issue exactly those commands to n3:
```js
const results = []
for (const operation of operations.split("\n")) {
  results.push(store.match(...JSON.parse(operation)));
}
```

____
#### Environment:

| software            | version
| ------------------- | -------
| Comunica Init Actor | 1.21.3 
| node                | v12.22.1
| npm                 | 6.14.12
| yarn                | 1.22.5
| Operating System    | linux (Linux 5.8.0-55-generic)

```bash
$ yarn why n3
yarn why v1.22.5
[1/4] Why do we have the module "n3"...?
[2/4] Initialising dependency graph...
[3/4] Finding dependency...
[4/4] Calculating file sizes...
=> Found "n3@1.10.0"
info Has been hoisted to "n3"
info Reasons this module exists
   - Specified in "dependencies"
   - Hoisted from "@comunica#actor-init-sparql-rdfjs#@comunica#actor-rdf-serialize-n3#n3"
   - Hoisted from "@comunica#actor-init-sparql-rdfjs#@comunica#actor-init-sparql#@comunica#actor-rdf-parse-n3#n3"
   - Hoisted from "@comunica#actor-init-sparql-rdfjs#@comunica#actor-init-sparql#@comunica#actor-query-operation-sparql-endpoint#fetch-sparql-endpoint#n3"
   - Hoisted from "@comunica#actor-init-sparql-rdfjs#@comunica#actor-init-sparql#@comunica#actor-rdf-resolve-hypermedia-none#rdf-store-stream#n3"
info Disk size without dependencies: "324KB"
info Disk size with unique dependencies: "540KB"
info Disk size with transitive dependencies: "708KB"
info Number of shared dependencies: 6
Done in 0.24s.
```

#### Crash log:

<details><summary>click here for OOM trace</summary>

```
$ node ./mwe.js 

<--- Last few GCs --->

[316734:0x3391e70]     6463 ms: Scavenge 2044.7 (2049.2) -> 2044.7 (2049.7) MB, 2.8 / 0.0 ms  (average mu = 0.109, current mu = 0.035) allocation failure 
[316734:0x3391e70]     6466 ms: Scavenge 2045.0 (2049.7) -> 2045.0 (2052.2) MB, 2.8 / 0.0 ms  (average mu = 0.109, current mu = 0.035) allocation failure 
[316734:0x3391e70]     6584 ms: Scavenge 2046.7 (2052.2) -> 2046.9 (2050.9) MB, 4.7 / 0.0 ms  (average mu = 0.109, current mu = 0.035) allocation failure 


<--- JS stacktrace --->

==== JS stack trace =========================================

    0: ExitFrame [pc: 0x140de99]
Security context: 0x2896f9cc08d1 <JSObject>
    1: addQuads [0x110e284d3b29] [/home/martin/issue-specific/5324/node_modules/n3/lib/N3Store.js:~282] [pc=0x2dee103d6272](this=0x118fdf11dd71 <N3Store map = 0x3e7aab97139>,0x35de594b0969 <JSArray[890]>)
    2: new constructor(aka N3Store) [0x110e284cfdd1] [/home/martin/issue-specific/5324/node_modules/n3/lib/N3Store.js:43] [bytecode=0x17475757ac59 offset=154](t...

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0xa1a640 node::Abort() [node]
 2: 0xa1aa4c node::OnFatalError(char const*, char const*) [node]
 3: 0xb9a68e v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [node]
 4: 0xb9aa09 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [node]
 5: 0xd57c85  [node]
 6: 0xd58316 v8::internal::Heap::RecomputeLimits(v8::internal::GarbageCollector) [node]
 7: 0xd64bd5 v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [node]
 8: 0xd65a85 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [node]
 9: 0xd6853c v8::internal::Heap::AllocateRawWithRetryOrFail(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [node]
10: 0xd2fc8e v8::internal::Factory::NewUninitializedFixedArray(int, v8::internal::AllocationType) [node]
11: 0xea06f1  [node]
12: 0xeb99f6  [node]
13: 0xeb9d3b  [node]
14: 0xf06932 v8::internal::JSObject::AddDataElement(v8::internal::Handle<v8::internal::JSObject>, unsigned int, v8::internal::Handle<v8::internal::Object>, v8::internal::PropertyAttributes) [node]
15: 0xf52eba v8::internal::Object::AddDataProperty(v8::internal::LookupIterator*, v8::internal::Handle<v8::internal::Object>, v8::internal::PropertyAttributes, v8::Maybe<v8::internal::ShouldThrow>, v8::internal::StoreOrigin) [node]
16: 0xf53bf7 v8::internal::Object::SetProperty(v8::internal::LookupIterator*, v8::internal::Handle<v8::internal::Object>, v8::internal::StoreOrigin, v8::Maybe<v8::internal::ShouldThrow>) [node]
17: 0x10882e4 v8::internal::Runtime::SetObjectProperty(v8::internal::Isolate*, v8::internal::Handle<v8::internal::Object>, v8::internal::Handle<v8::internal::Object>, v8::internal::Handle<v8::internal::Object>, v8::internal::StoreOrigin, v8::Maybe<v8::internal::ShouldThrow>) [node]
18: 0x10894aa v8::internal::Runtime_SetKeyedProperty(int, unsigned long*, v8::internal::Isolate*) [node]
19: 0x140de99  [node]
Aborted (core dumped)
```
</details>

**Comment by rubensworks** [comment]
I would expect our current sequence property path implementation to be sufficiently efficient. However, the combination with the filter and other triple pattern may result in a very expensive join, which also lies at the root cause of #725.

Several people have been running into similar problems lately, so we should urgently think about a proper way forward for this.

**Comment by rubensworks** [comment]
I just tried reproducing this myself in the latest dev version, and I don't seem to be running into any memory issues.
I therefore suspect that this has been resolved because of #826.
So this should be fixed in the next release.

Feel free to try out the current dev version by cloning from master.
I've managed to execute your snippet with just 64 MB of memory.
